### ISSUE TITLE
Proposal on improving interactive_shell usability (seen in build and debrew)

### ISSUE THREAD

**Issue by zmwangx** [Homebrew/homebrew#48470]
**TL; DR.** Skip the next paragraph, which is background plus complaints.

---

Let's admit it, the `interactive_shell` used by `build` or `debrew` is pretty hostile (make no mistake — having the ability to open an interactive shell when in need is awesome; but the usability of the shell is usually less than satisfactory). Most of us customize our shells to some extent, and some of us (I, for one) customize very heavily. However, during formula installation `$HOME` is set to `buildpath/".brew_home"`, which is good for isolation but not so good for an interactive shell, where everything just falls back to defaults and immediately feels like stone age — aliases etched into muscle memory not present, familiar shell options not set, completion half broken, no dircolors from any sort of listings, etc. (I guess fish shell won't have this problem, but that's not my thing.) And I've been annoyed by this hostile shell environment  enough times to raise this issue.

I think this problem could be solved by allowing the user to specify an environment preparation script, which is run before actually starting the shell. Then for users with light customizations, a simple script that symlinks a `.bashrc` or `.zshrc` might be enough; for me I would copy over my entire `.config` directory and run a readily-baked setup script:

```zsh
#!/usr/bin/env zsh
rsync -a ~$USER/.config --exclude='*.git' ~
~/.config/setup 2>/dev/null
```

but the point is everyone could customize to their needs.

To make it concrete, here's a potential patch to implement my proposed solution:

```diff
diff --git a/Library/Homebrew/utils.rb b/Library/Homebrew/utils.rb
index d9a41b8..7dea0cc 100644
--- a/Library/Homebrew/utils.rb
+++ b/Library/Homebrew/utils.rb
@@ -156,13 +156,22 @@ def plural(n, s = "s")
   (n == 1) ? "" : s
 end
 
+def homedir_is_temporary()
+  ENV["HOME"].start_with?(HOMEBREW_TEMP.resolved_path.to_s)
+end
+
 def interactive_shell(f = nil)
   unless f.nil?
     ENV["HOMEBREW_DEBUG_PREFIX"] = f.prefix
     ENV["HOMEBREW_DEBUG_INSTALL"] = f.full_name
   end
 
-  if ENV["SHELL"].include?("zsh") && ENV["HOME"].start_with?(HOMEBREW_TEMP.resolved_path.to_s)
+  env_prep_script = ENV["HOMEBREW_ENV_PREPARE"]
+  if homedir_is_temporary && env_prep_script && File.executable?(env_prep_script)
+    system env_prep_script
+  end
+
+  if homedir_is_temporary && ENV["SHELL"].include?("zsh")
     FileUtils.touch "#{ENV["HOME"]}/.zshrc"
   end
 
```

Here I'm proposing an environment variable `HOMEBREW_ENV_PREPARE` (or something like that) for specifying the environment preparation script. I realize that throwing in a random env var for every random problem is bad, but I can't see another way around right now.

Not sure if anyone would be interested in this problem or proposal.

**Comment by mikemcquaid** [comment]
I agree the interactive environment is annoying to work with. I think we could get away with just doing `PATH` modification by adding e.g. superenv and dependencies to the beginning of the PATH, removing `HOMEBREW_PREFIX` (and things inside it) from the `PATH` and not overriding `HOME`. While that would make it less like the actual Homebrew environment I think it's still probably worth it.

**Comment by zmwangx** [comment]
That would be another good solution.

**Comment by mikemcquaid** [comment]
@zmwangx Could you try and open a pull request? This should help and I'm happy to walk you through anything else: https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/How-To-Open-a-Homebrew-Pull-Request-(and-get-it-merged).md#how-to-open-a-homebrew-pull-request-and-get-it-merged

Thanks!
